lib/src/deploy.rs

```diff
@@ -301,15 +301,41 @@ async fn handle_layer_progress_print(
     prog
 }
 
-/// Wrapper for pulling a container image, wiring up status output.
-#[context("Pulling")]
-pub(crate) async fn pull(
+/// Gather all bound images in all deployments, then prune the image store,
+/// using the gathered images as the roots (that will not be GC'd).
+pub(crate) async fn prune_container_store(sysroot: &Storage) -> Result<()> {
+    let deployments = sysroot.deployments();
+    let mut all_bound_images = Vec::new();
+    for deployment in deployments {
+        let bound = crate::boundimage::query_bound_images_for_deployment(sysroot, &deployment)?;
+        all_bound_images.extend(bound.into_iter());
+    }
+    // Convert to a hashset of just the image names
+    let image_names = HashSet::from_iter(all_bound_images.iter().map(|img| img.image.as_str()));
+    let pruned = sysroot
+        .get_ensure_imgstore()?
+        .prune_except_roots(&image_names)
+        .await?;
+    tracing::debug!("Pruned images: {}", pruned.len());
+    Ok(())
+}
+
+pub(crate) struct PreparedImagePull {
+    pub imp: ImageImporter,
+    pub prep: Box<PreparedImport>,
+    pub digest: Digest,
+    pub n_layers_to_fetch: usize,
+    pub layers_total: usize,
+    pub bytes_to_fetch: u64,
+    pub bytes_total: u64,
+}
+
+#[context("Prepare for pull")]
+pub(crate) async fn prepare_for_pull(
     repo: &ostree::Repo,
     imgref: &ImageReference,
     target_imgref: Option<&OstreeImageReference>,
-    quiet: bool,
-    prog: ProgressWriter,
-) -> Result<Box<ImageState>> {
+) -> Result<(Option<PreparedImagePull>, Option<Box<ImageState>>)> {
```

Comment:
We discussed converting this to an `enum`; maybe we can either just return the existing one (maybe?) or make a new thin wrapper like this

```
enum PullPreparedState {
   Ready(PreparedImagePull)
   AlreadyPresent(Box<ImageState>)
}
```

Reply:
switched this to an enum. I had to use a wrapper to at least include the `ImageImporter` and `Box<PreparedImport>` in the response.